File: greedy/1282_group_the_people_given_the_group_size_they_belong_to/1282_group_the_people_given_the_group_size_they_belong_to.go
package main

import "fmt"

func groupThePeople(groupSizes []int) [][]int {
	var groups [][]int
	groupMemberCount := make(map[int][]int)

	for i := 0; i < len(groupSizes); i++ {
		groupMemberCount[groupSizes[i]] = append(groupMemberCount[groupSizes[i]], i)
	}

	for groupSize, groupMembers := range groupMemberCount {
		subGroup := []int{}
		for i := 0; i < len(groupMembers); i++ {
			if i%groupSize == groupSize-1 {
				subGroup = append(subGroup, groupMembers[i])
				groups = append(groups, subGroup)
				subGroup = []int{}
				continue
			}
			subGroup = append(subGroup, groupMembers[i])
		}
	}
	return groups
}

func main() {
	groupSizes := []int{3, 3, 3, 3, 3, 1, 3}
	res := groupThePeople(groupSizes)
	fmt.Println(res)
}

File: greedy/1282_group_the_people_given_the_group_size_they_belong_to/1282_group_the_people_given_the_group_size_they_belong_to.py
import collections

from typing import List

class Solution:
    def groupThePeople(self, groupSizes: List[int]) -> List[List[int]]:
        groups = collections.defaultdict(list)
        for i, _id in enumerate(groupSizes):
            groups[_id].append(i)
        
        ans = list()
        for gsize, users in groups.items():
            for it in range(0, len(users), gsize):
                ans.append(users[it : it + gsize])
        return ans


groupSizes = [3,3,3,3,3,1,3]
res = Solution().groupThePeople(groupSizes)
print(res)